dbms/src/Common/BitpackingPrimitives.h

```diff
@@ -0,0 +1,340 @@
+    template <typename T, bool is_signed = std::numeric_limits<T>::is_signed>
+    constexpr static UInt8 minimumBitWidth(T value)
+    {
+        return findMinimumBitWidth<T, is_signed, BYTE_ALIGNED>(value, value);
+    }
+
+    // Calculates the minimum required number of bits per value that can store all values
+    template <typename T, bool is_signed = std::numeric_limits<T>::is_signed>
+    constexpr static UInt8 minimumBitWidth(const T * values, size_t count)
+    {
+        return findMinimumBitWidth<T, is_signed, BYTE_ALIGNED>(values, count);
+    }
+
+    // Calculates the minimum required number of bits per value that can store all values,
+    // given a predetermined minimum and maximum value of the buffer
+    template <typename T, bool is_signed = std::numeric_limits<T>::is_signed>
+    constexpr static UInt8 minimumBitWidth(T minimum, T maximum)
+    {
+        return findMinimumBitWidth<T, is_signed, BYTE_ALIGNED>(minimum, maximum);
+    }
+
+    constexpr static size_t getRequiredSize(size_t count, UInt8 width)
+    {
+        count = roundUpToAlgorithmGroupSize(count);
+        return ((count * width) / 8);
+    }
+
+    template <typename T>
+    constexpr static T roundUpToAlgorithmGroupSize(T num_to_round)
+    {
+        int remainder = num_to_round % BITPACKING_ALGORITHM_GROUP_SIZE;
+        if (remainder == 0)
+        {
+            return num_to_round;
+        }
+
+        return num_to_round + BITPACKING_ALGORITHM_GROUP_SIZE - remainder;
+    }
+
+private:
+    template <typename T, bool is_signed, bool round_to_next_byte = false>
+    constexpr static UInt8 findMinimumBitWidth(const T * values, size_t count)
+    {
+        T min_value = values[0];
+        T max_value = values[0];
+
+        for (size_t i = 1; i < count; i++)
+        {
+            if (values[i] > max_value)
+            {
+                max_value = values[i];
+            }
+
+            if constexpr (is_signed)
+            {
+                if (values[i] < min_value)
+                {
+                    min_value = values[i];
+                }
+            }
+        }
+        return findMinimumBitWidth<T, is_signed, round_to_next_byte>(min_value, max_value);
+    }
+
+    template <typename T, bool is_signed, bool round_to_next_byte = false>
+    constexpr static UInt8 findMinimumBitWidth(T min_value, T max_value)
+    {
+        UInt8 bitwidth;
+        T value;
+
+        if constexpr (is_signed)
+        {
+            if (min_value == std::numeric_limits<T>::min())
+            {
+                // handle special case of the minimal value, as it cannot be negated like all other values.
+                return sizeof(T) * 8;
+            }
+            else
+            {
+                value = std::max(static_cast<T>(-min_value), max_value);
+            }
+        }
+        else
+        {
+            value = max_value;
+        }
+
+        if (value == 0)
```

Comment:
Could it be moved before the `is_signed` test? signed and unsigned zero seems make no difference here

Reply:
```cpp
        T value;
        if (value == 0)
        {
            return 0;
        }

        if constexpr (is_signed)
        {
            if (min_value == std::numeric_limits<T>::min())
            {
                // handle special case of the minimal value, as it cannot be negated like all other values.
                return sizeof(T) * 8;
            }
            else
            {
                value = std::max(static_cast<T>(-min_value), max_value);
            }
        }
        else
        {
            value = max_value;
        }
```

like this?

Reply:
sorry, not get your point

Reply:
I wonder why don't we put it at the very beginning?

Reply:
`value` is not initialize

Reply:
got